cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/AppFabricTestBase.java

```diff
@@ -539,21 +539,36 @@ protected HttpResponse addArtifact(Id.Artifact artifactId, InputSupplier<? exten
     return addArtifact(artifactId, artifactContents, parents, Collections.emptySet());
   }
 
-
   protected HttpResponse addArtifact(Id.Artifact artifactId, InputSupplier<? extends InputStream> artifactContents,
                                      Set<ArtifactRange> parents, Set<PluginClass> pluginClasses) throws Exception {
+    if (pluginClasses != null && !pluginClasses.isEmpty()) {
+      return addArtifactHelper(artifactId, artifactContents, parents, GSON.toJson(pluginClasses));
+    } else {
+      return addArtifactHelper(artifactId, artifactContents, parents, null);
+    }
+  }
+
+  /**
+   * This method is a helper method which accepts GSON serialized form of plugin classes for testing purpose only.
+   * It's usage is discouraged. Please consider using {@link #addArtifact(Id.Artifact, InputSupplier, Set, Set)}
+   */
+  protected HttpResponse addArtifactHelper(Id.Artifact artifactId,
+                                           InputSupplier<? extends InputStream> artifactContents,
+                                           Set<ArtifactRange> parents, @Nullable String pluginClassesJSON)
+    throws Exception {
     String path = getVersionedAPIPath("artifacts/" + artifactId.getName(), artifactId.getNamespace().getId());
     HttpEntityEnclosingRequestBase request = getPost(path);
     request.setHeader(Constants.Gateway.API_KEY, "api-key-example");
     request.setHeader("Artifact-Version", artifactId.getVersion().getVersion());
     if (parents != null && !parents.isEmpty()) {
       request.setHeader("Artifact-Extends", Joiner.on('/').join(parents));
     }
-
-    if (pluginClasses != null && !pluginClasses.isEmpty()) {
-      request.setHeader("Artifact-Plugins", GSON.toJson(pluginClasses));
+    // Note: we purposefully do no check for empty string and let it pass as the header because we want to test the
+    // behavior where plugin classes header is set to empty string. This is what is passed by the UI. For more
+    // details see: https://issues.cask.co/browse/CDAP-14578
+    if (pluginClassesJSON != null) {
```

Comment:
The jira says issue happens when header is set as empty string. In case `pluginClassesJSON ` is empty, should we set header `Artifact-Plugins` as empty string?

Reply:
Don't like the idea setting stuff to null. Also if we do we depend on setHeader to handle null for us currently they skip set if value is null which might change in future causing our code to behave differently. It will be better handle it ourselves. 

Also the check is happening whether we do it or not. By doing early we save a function call.

Reply:
nvm, the `pluginClassesJSON` is empty in the tests.